src/braket/default_simulator/openqasm/data_manipulation.py

```diff
@@ -0,0 +1,656 @@
+    "exp": lambda x: FloatLiteral(np.exp(x.value)),
+    "floor": lambda x: IntegerLiteral(math.floor(x.value)),
+    "log": lambda x: FloatLiteral(np.log(x.value)),
+    "mod": lambda x, y: (
+        IntegerLiteral(x.value % y.value)
+        if isinstance(x, IntegerLiteral) and isinstance(y, IntegerLiteral)
+        else FloatLiteral(x.value % y.value)
+    ),
+    "popcount": lambda x: popcount(x),
+    # parser gets confused by pow, mistaking for quantum modifier
+    "pow": lambda x, y: (
+        IntegerLiteral(x.value**y.value)
+        if isinstance(x, IntegerLiteral) and isinstance(y, IntegerLiteral)
+        else FloatLiteral(x.value**y.value)
+    ),
+    "rotl": lambda x: NotImplementedError(),
+    "rotr": lambda x: NotImplementedError(),
+    "sin": lambda x: FloatLiteral(np.sin(x.value)),
+    "sqrt": lambda x: FloatLiteral(np.sqrt(x.value)),
+    "tan": lambda x: FloatLiteral(np.tan(x.value)),
+}
+
+
+def resolve_type_hierarchy(x: Expression, y: Expression):
+    """Determine output type of expression, for example: 1 + 1.0 == 2.0"""
+    return max(type(x), type(y), key=type_hierarchy.index)
+
+
+def evaluate_binary_expression(lhs: Expression, rhs: Expression, op: BinaryOperator):
+    result_type = resolve_type_hierarchy(lhs, rhs)
+    func = operator_maps[result_type].get(op)
+    if not func:
+        raise TypeError(f"Invalid operator {op.name} for {result_type.__name__}")
+    return func(lhs, rhs)
+
+
+def evaluate_unary_expression(expression: Expression, op: BinaryOperator):
+    expression_type = type(expression)
+    func = operator_maps[expression_type].get(op)
+    if not func:
+        raise TypeError(f"Invalid operator {op.name} for {expression_type.__name__}")
+    return func(expression)
+
+
+def get_operator_of_assignment_operator(assignment_operator: AssignmentOperator):
+    return getattr(BinaryOperator, assignment_operator.name[:-1])
+
+
+"""
+Casting values
+"""
+
+
+LiteralType = Union[BooleanLiteral, IntegerLiteral, FloatLiteral, ArrayLiteral, BitstringLiteral]
+
+
+@singledispatch
+def cast_to(into: Union[ClassicalType, Type[LiteralType]], variable: LiteralType):
+    """Cast a variable into a given type. Order of parameters is to enable singledispatch"""
+    if type(variable) == into:
+        return variable
+    if into == BooleanLiteral or isinstance(into, BoolType):
+        return BooleanLiteral(bool(variable.value))
+    if into == IntegerLiteral:
+        return IntegerLiteral(int(variable.value))
+    if into == FloatLiteral:
+        return FloatLiteral(float(variable.value))
+    raise TypeError(f"Cannot cast {type(variable).__name__} into {into.__name__}.")
+
+
+@cast_to.register
+def _(into: BitType, variable: Union[BooleanLiteral, ArrayLiteral, BitstringLiteral]):
+    """
+    Bit types can be sized or not, represented as Boolean literals or Array literals.
+    Sized bit types can be instantiated with a Bitstring literal or Array literal.
+    """
+    if not into.size:
```

Comment:
Is this the same as checking if isinstance(variable, BooleanLiteral), or not?

Reply:
You are correct in that that is when that branch happens. The slight difference would be in error handling. I think the logic is a little cleaner if split by intent than by type of thing being cast. But yes with perfect code, those should be equivalent